CAM: snapmaker improve support for different machine variants and toolheads

jalapenopuzzle opened the pull request:
This PR attempts to support all of the Snapmaker machines and their variants and modifications.
For machines with only one compatible toolhead, the toolhead is selected automatically.
For machines supporting multiple compatible toolheads, the --toolhead option is required to select the appropriate one.
The names of the toolheads have been expanded to include the key _CNC, to allow for possible future support of the various laser and 3DP toolheads available on the Snapmaker platform.
This patch addresses a defect in support for the 200W CNC toolhead (RPM speed option), where the gcode was not returned from convert_spindle(), causing strange issues in export_common().

There are several topics on my TODO list which may be the subject of future PRs:
* Limit feed rates in the various cartesian directions, particularly the Z direction.
Z feed rates above 600mm/min make my machine lose steps and cause "barnacle" layers in 3D prints.
* G2/G3 are implemented in the Snapmaker firmware but are not well tested
(according to Snapmaker's GCode documentation).
Snapmaker recommends replacing G2/G3 with interpolated G0/G1 moves as appropriate.
I am not sure how to approach this issue in the refactored post-processing implementation.
* Checking feed rates and spindle speeds are greater than 0 and raising errors as appropriate.

I have an earlier implementation before the CAM post-processing refactoring which addressed these topics. I have not figured out how to port that code into the current main branch given the refactoring. I figured I should raise this PR to begin the process of getting these modest changes merged before attempting the other topics.

I am not sure how to run unit tests on this code yet. Please school me gently.

I have only attempted to test this on my own machine, a Snapmaker A350T with 200W CNC toolhead.
I do also have the 50W toolhead but I haven't attempted to test with it.
I am hoping others in the Snapmaker community can review this work and contribute fixes as necessary, particularly for Snapmaker models which I do not have.

I am finding that the Snapmaker is very limited in it's CNC capabilities, and I keep finding that the machine is not rigid enough to do what I want it to do. This is making testing difficult for me.

LarryWoestman commented:
As far as I can tell, the refactored code base does not have the ability to convert G2/G3 to G0/G1 moves.

https://forum.freecad.org/viewtopic.php?t=95704 has some info on how to run some of the existing tests for postprocessors.  The "refactored*.py" files in src/Mod/CAM/CAMTests contain some examples of how to write tests to verify if a postprocessor is outputting the desired g-code.  

https://forum.freecad.org/viewtopic.php?t=95054 might also be worth reading.

phaseloop commented:
Feed rate limiting and loosing steps is and should be a matter of firmware and controller configuration. I don't have an experience with Snapmaker machines but they claim they are based on Marlin firmware (this also explains partial support for G2/G3). Max speed and acceleration for all axes are set using M203 g-code (+ M500 to save if I remember).

https://marlinfw.org/docs/gcode/M203.html

As for detection of spindle speed 0 - I suppose it would make sense to keep speed 0 if you use laser. I'm working on fixing a mess within toolbit controller. There used to be a ToolType attribute which was removed but some operations and postprocessors (like fanuc postprocessor or surface operation) still depend on it (or use awful hacks to infer tool type). 

So I plan to bring back ToolType into bit configuration and in the long term also add laser cutting/engraving operations in CAM.

LarryWoestman commented:
I recall one discussion where we talked about the possibility of replacing the "command line arguments" for postprocessors with an appropriate GUI.  Nice idea that will be implemented when we have lots of CAM programmers looking for something useful to do :-) , or (more likely) when a CAM developer decides to take on such a large task.

phaseloop commented:
I know this is a theoretical discussion - but I wasn't thinking about replacing command line arguments - just using additional GUI widgets to create those - at least parse and show user what options are possible to use.

jalapenopuzzle commented on src/Mod/CAM/Path/Post/scripts/snapmaker_post.py:
```diff
@@ -286,21 +332,36 @@ def snapmaker_init_parser(
             default=None,
             choices=self.values["TOOLHEADS"].keys(),
-            help=f"Snapmaker toolhead",
+            help=f"Snapmaker toolhead. Choose from [{self.values['TOOLHEADS'].keys()}].",
```
I think this is a great idea that belongs in a new ticket...
The `argparse` options can be processed and queried.
These options are returned by `snapmaker_init_parser()`, which makes them accessible for extraction and adding to the GUI. But there are a lot of options, and the trick will be to filter which ones should be presented on the GUI.

jalapenopuzzle commented on src/Mod/CAM/Path/Post/scripts/snapmaker_post.py:
```diff
@@ -339,46 +394,92 @@ def snapmaker_init_parser(
+                        self.values["BOUNDARIES"]["Z"] -= 15
+                if args.bracing_kit:
+                    if not self.values["MACHINE_NAME"].startswith("Snapmaker 2 "):
+                        FreeCAD.Console.PrintError(
```
I have moved the mod kits {bracing kit, quick swap kit} into their own separate configuration dict and made the option processing generic based on that dict. The kit dimensions have been moved into the configuration dict. The installed kit names are now accumulated into a list MOD_KITS_INSTALLED.

jalapenopuzzle commented on src/Mod/CAM/Path/Post/scripts/snapmaker_post.py:
```diff
@@ -46,17 +46,71 @@
+        compatible_toolheads={"Original_CNC"},
+        lead=dict(X=8, Y=8, Z=8),  # Linear module screw pitch (mm/turn)
+    ),
+    a150=dict(
```
I have added a change to capitalise all of the machine names.

sliptonic commented:
It should still be possible to script the post-processing and gcode generation.  We want to retain that possibility so that CAM can be included in automated workflows.  This need not be done through a command line but it should NOT assume that a human being is involved.

LarryWoestman commented:
Agree that postprocessing and g-code generation must remain easy to do from a program.

jalapenopuzzle commented:
I have worked out how to run the unit tests and how to extend them.
I have also re-worked almost every commit on the branch so that each commit passes the unit tests.
I have only tested on Ubuntu 24.04, but I see that the github workflow will test on lots of platforms that I could not.
Each commit that adds new functionality also adds appropriate unit tests to ensure that the new functionality works as intended.
I believe I have addressed the review comments, with the exception of inventing a new option passing and interrogation system to support GUI controls. That work belongs in a separate branch.
I have not addressed the G2/G3 issues yet. It seems that the core does not support re-writing G2/G3 as G1 moves. I will seek advice separately about the best approach for this issue.

LarryWoestman commented:
I agree that handling options with GUI belongs in a separate PR (or set of PR's), eventually.

Rewriting G2/G3 as G1 moves should also be a separate PR (or set of PR's).  If you or someone else writes the code to do this I would be happy to put it in the refactored code base so it is available for use in all of the refactored postprocessors.  There may be tradeoffs related to gcode size but it would still be a nice option to have (someday).

jalapenopuzzle commented:
Branch rebased without changes, unit tested (CAM only - pass) and force pushed.

jalapenopuzzle commented:
I have rebased, re-tested (TestCAMApp tests all OK on Ubuntu 24.04), and force pushed again.
The github workflows need to be approved.
I can't approve them.
How can we move forward with getting this closer to merge?

sliptonic commented:
started the workflow

jalapenopuzzle commented:
@sliptonic Thanks for running the workflow.
What's the next step towards getting this merged?
Every time I rebase it, I need you to run the tests again, but then nothing happens and the branch gets stale again.
Anyway, I can rebase and re-test on my own machine (Ubuntu 24.04) any time.

@LarryWoestman I did add G2/G3 translation in a separate PR but it's only a draft for now.
<https://github.com/FreeCAD/FreeCAD/pull/20981>

@phaseloop the mod kits for the Snapmaker 2.0 series were developed to address shortcomings in the design of the machine (developed as a kickstarter campaign). Even with the bracing kit, the machine is not rigid enough for interesting CNC work so a lot of people do custom mods to stiffen it up. And then Snapmaker released the Artisan, which is built more heavily and has no mod kits.
In the end I gave up trying to generalise the size changes introduced by each mod kit, as the working volume depends on the toolhead mount offset and how that interacts with both sides of the working envelope, for each dimension. There was no generalisation to be had without reverse engineering the entire computation including toolhead mounting offsets. So I have used the manufacturer's published build volume table for each supported combination of mod kits and tool heads. If a user needs to do something weird corresponding to a custom modification, it is possible to define your own custom bounds and bypass setting the mod kits entirely.

I did add a fairly comprehensive set of unit tests for this work, and everything passes.

The place the reviewers might complain is that the sequence of commits is long and I did change my approach to solving the boundary problem part way through. I have not attempted to re-work the earlier commits to avoid the change in approach.

Snapmaker also has a rotary attachment for Snapmaker 2.0 and Artisan. I have one, and I may attempt to add support for it in the future (but I'm not sure if FreeCAD supports generating tool paths for 4 axis CNC machines). This is a topic for a separate PR.

LarryWoestman commented:
@jalapenopuzzle Thanks, I saw the G2/G3 draft PR.  I hope to have time tomorrow to look at in detail.

The "usual" release process expects to have all of the commits "squashed" into one commit.  Otherwise it is more difficult or not possible to use the current process.  I can give you instructions and/or help you figure out how to do this if you haven't done it before.

I normally try to keep up with the "main" branch at intervals of around a week or so.  If the PR gets too far behind "main" the chances that the PR will merge without issues goes down.

LarryWoestman commented:
@chennes I need to do another careful reading of this PR to see if I can spot anything worth worrying about, given that it has had some changes since my last reading.

LarryWoestman reviewed (commented):
In general this looks reasonable.  I don't know enough about Snapmaker to know if it is "correct".

I'm not familiar with the license you are using.  Is it "compatible" with our usual license (LGPL2+) or does it have extra "requirements"?
